### PR TITLE
Fix handling of object versions pending purge

### PR DESCRIPTION
**Pull request by krisis**
## Description
- GetObject() with vid should return 405
- GetObject() without vid should return 404
- ListObjects() should ignore this object if this is the "latest" version of the object
- ListObjectVersions() should list this object as "DELETE marker"
- Remove data parts before sync'ing the version pending purge

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation updated
- [ ] Unit tests added/updated

### CODE REVIEW

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

File: cmd/xl-storage-format-v2.go [hunk 1]
```diff
@@ -595,6 +595,9 @@ func (j xlMetaV2Object) ToFileInfo(volume, path string) (FileInfo, error) {
 		}
 	}
 	fi.ReplicationState = getInternalReplicationState(fi.Metadata)
+	if !fi.VersionPurgeStatus().Empty() {
+		fi.Deleted = true
+	}
```
Comment (harshavardhana) [inline cmd/xl-storage-format-v2.go]:
```suggestion
	fi.Deleted =  !fi.VersionPurgeStatus().Empty()
```

File: cmd/erasure-object.go [hunk 2]
```diff
@@ -493,9 +501,12 @@ func (er erasureObjects) getObjectInfo(ctx context.Context, bucket, object strin
...
 	objInfo = fi.ToObjectInfo(bucket, object)
-	if opts.VersionID != "" && !fi.VersionPurgeStatus().Empty() {
-		// Make sure to return object info to provide extra information.
-		return objInfo, toObjectErr(errMethodNotAllowed, bucket, object)
+	if !fi.VersionPurgeStatus().Empty() {
+		if opts.VersionID != "" {
+			// Make sure to return object info to provide extra information.
+			return objInfo, toObjectErr(errMethodNotAllowed, bucket, object)
+		}
+		return objInfo, toObjectErr(errFileNotFound, bucket, object)
```
Comment (harshavardhana) [inline cmd/erasure-object.go]: not needed since fi.deleted is set to true

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

File: cmd/erasure-object.go [hunk 3]
```diff
@@ -204,6 +204,14 @@ func (er erasureObjects) GetObjectNInfo(ctx context.Context, bucket, object stri
 		}
 	}
 
+	if !fi.VersionPurgeStatus().Empty() {
+		if opts.VersionID != "" {
+			// Make sure to return object info to provide extra information.
+			return nil, toObjectErr(errMethodNotAllowed, bucket, object)
+		}
+		return nil, toObjectErr(errFileNotFound, bucket, object)
+	}
```
Comment (harshavardhana) [inline cmd/erasure-object.go]: This is not needed as well @krisis

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Review by harshavardhana — APPROVED** [review]
This addresses all the things that we thought it would.

**Review by poornas — APPROVED** [review]
LGTM

**Pull request merged**